packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,602 @@
+///                             },
+///                           ),
+///                         ],
+///                       );
+///                     },
+///                   );
+///                 },
+///                 child: Text('Submit'),
+///               ),
+///             ],
+///           ),
+///         ),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+class AutocompleteCore<T extends Object> extends StatefulWidget {
+  /// Create an instance of AutocompleteCore.
+  ///
+  /// [fieldViewBuilder] and [optionsViewBuilder] must not be null.
+  const AutocompleteCore({
+    Key? key,
+    required this.fieldViewBuilder,
+    required this.optionsViewBuilder,
+    required this.optionsBuilder,
+    this.displayStringForOption = _defaultStringForOption,
+    this.onSelected,
+  }) : assert(displayStringForOption != null),
+       assert(fieldViewBuilder != null),
+       assert(optionsBuilder != null),
+       assert(optionsViewBuilder != null),
+       super(key: key);
+
+  /// Builds the field whose input is used to get the options.
+  ///
+  /// Pass the provided [TextEditingController] to the field built here so that
+  /// AutocompleteCore can listen for changes.
+  final AutocompleteFieldViewBuilder fieldViewBuilder;
+
+  /// Builds the selectable options widgets from a list of options objects.
+  ///
+  /// The options are displayed floating below the field using a
+  /// [CompositedTransformFollower] inside of an [Overlay], not at the same
+  /// place in the widget tree as AutocompleteCore.
+  final AutocompleteOptionsViewBuilder<T> optionsViewBuilder;
+
+  /// Returns the string to display in the field when the option is selected.
+  ///
+  /// This is useful when using a custom T type and the string to display is
+  /// different than the string to search by.
+  ///
+  /// If not provided, will use `option.toString()`.
+  final AutocompleteOptionToString<T> displayStringForOption;
+
+  /// Called when an option is selected by the user.
+  ///
+  /// Any [TextEditingController] listeners will not be called when the user
+  /// selects an option, even though the field will update with the selected
+  /// value, so use this to be informed of selection.
+  final AutocompleteOnSelected<T>? onSelected;
+
+  /// A function that returns the current selectable options objects given the
+  /// current TextEditingValue.
+  final AutocompleteOptionsBuilder<T> optionsBuilder;
+
+  // The default way to convert an option to a string.
+  static String _defaultStringForOption(dynamic option) {
+    return option.toString();
+  }
+
+  @override
+  _AutocompleteCoreState<T> createState() => _AutocompleteCoreState<T>();
+}
+
+class _AutocompleteCoreState<T extends Object> extends State<AutocompleteCore<T>> {
+  final GlobalKey _fieldKey = GlobalKey();
+  final LayerLink _optionsLayerLink = LayerLink();
+  final TextEditingController _textEditingController = TextEditingController();
```

Comment:
Should we let the user bring their own controller?

Reply:
Yes, that is my next PR!  I tried to simplify it by keeping it private for now.  It will be necessary to use your own controller if you want to do things like split the field and results into different parts of the widget tree.  I'll be tagging you for review soon :)